src/reducers/vis-state-updaters.js

```diff
@@ -585,23 +587,30 @@ export function setFilterUpdater(state, action) {
       // only filter the current dataset
       break;
     case FILTER_UPDATER_PROPS.layerId:
-      // const layers = state.layers.filter(l => value.includes(l.id));
+      // We need to update only datasetId/s if we have added/removed layers
+      // - check for layerId changes (XOR works because of string values)
+      // if no differences between layerIds, don't do any filtering
+      const layerIdDifference = xor(newFilter.layerId, oldFilter.layerId);
 
       const layerDataIds = uniq(
-        value
+        layerIdDifference
           .map(lid => get(state.layers.find(l => l.id === lid), ['config', 'dataId']))
           .filter(d => d)
       );
 
-      // diff filter.dataId with layerDataIds, to determine which dataset to run filter on
-      datasetIds = [
-        ...layerDataIds.filter(lid => !datasetIds.includes(lid)),
-        ...datasetIds.filter(did => !layerDataIds.includes(did))
-      ];
+      // only filter datasetsIds
+      datasetIds = layerDataIds;
+
+      // Update newFilter dataIds
+      const newDataIds = uniq(
```

Comment:
simply use `layerDataIds ` seems sufficient, why do you have to recalculate `newDataIds`?

Reply:
LayerDataIds is used to determine which datasets to perform the filter against.
The use case is the following:
- add a layer A (dataset_1) to the filter:
  - layerDataIds: [dataset_1]
  - newDataIds: [dataset_1]

- add a layer B (dataset_2) to the filter:
  - layerDataIds: [dataset_2]
  - newDataIds: [dataset_1, dataset_2]

- add a layer C (dataset_1) to the filter:
  - layerDataIds: [dataset_1]
  - newDataIds: [dataset_1, dataset_2]

__layerDataIds__ is necessary to determine which dataset to run the filter against during the current step.
__newDataIds__ is required to update the filter object with. the list of unique dataset ids